# House prices
    - null

In [2]:
import pandas as pd
import numpy as np


In [3]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')
full_data = [train,test]

In [4]:
train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [5]:
lst =np.array(train.isnull().sum())
# PoolQC           1453
# Fence            1179
# MiscFeature      1406
# Alley            1369
lst

# 결측치 200개 이상 컬럼 삭제
for i in range(2) :
    full_data[i]= full_data[i].drop(['Id','LotFrontage','PoolQC','FireplaceQu','Fence','MiscFeature','Alley'], axis = 1)

In [6]:

# 나머지 결측 행 삭제
for i in range(2):
    full_data[i] =  full_data[i].dropna(axis=0)
lst = np.array(full_data[1].isnull().sum())
lst

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0], dtype=int64)

# 가변수화

In [7]:

for i in range(2):
    full_data[i] = pd.get_dummies(full_data[i], drop_first =True)
full_data[1].head()

MSSubClass  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
0          20    11622            5            6       1961          1961   
1          20    14267            6            6       1958          1958   
2          60    13830            5            5       1997          1998   
3          60     9978            6            6       1998          1998   
4         120     5005            8            5       1992          1992   

   MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF          ...            \
0         0.0       468.0       144.0      270.0          ...             
1       108.0       923.0         0.0      406.0          ...             
2         0.0       791.0         0.0      137.0          ...             
3        20.0       602.0         0.0      324.0          ...             
4         0.0       263.0         0.0     1017.0          ...             

   SaleType_ConLI  SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0               0               0             0             0            1   
1               0               0             0             0            1   
2               0               0             0             0            1   
3               0               0             0             0            1   
4               0               0             0             0            1   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                      0                     0                     0   
1                      0                     0                     0   
2                      0                     0                     0   
3                      0                     0                     0   
4                      0                     0                     0   

   SaleCondition_Normal  SaleCondition_Partial  
0                     1                      0  
1                     1                      0  
2                     1                      0  
3                     1                      0  
4                     1                      0  

[5 rows x 206 columns]

# X,Y split

In [8]:
from sklearn.model_selection import train_test_split

tra , val = train_test_split(full_data[0],test_size = 0.2, random_state = 22)
x_train, y_train = tra.drop('SalePrice', axis = 1), tra.loc[:,'SalePrice']
x_val,y_val = val.drop('SalePrice', axis = 1), val.loc[:,'SalePrice']
# x_test, y_test = full_data[1].drop('SalePrice', axis = 1), full_data[1].iloc[:,'SalePrice']
x_train.head()


MSSubClass  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
465         120     3072            7            5       2004          2004   
924          20    10240            6            6       1980          1980   
23          120     4224            5            7       1976          1976   
422          20    21750            5            5       1954          1954   
610          60    11050            9            5       2000          2000   

     MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF          ...            \
465        18.0           0           0       1375          ...             
924       157.0         625        1061          0          ...             
23          0.0         840           0        200          ...             
422         0.0           0           0        988          ...             
610       204.0         904           0        536          ...             

     SaleType_ConLI  SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
465               0               0             0             0            1   
924               0               0             0             0            1   
23                0               0             0             0            1   
422               0               0             0             0            1   
610               0               0             0             0            1   

     SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
465                      0                     0                     0   
924                      0                     0                     0   
23                       0                     0                     0   
422                      0                     0                     0   
610                      0                     0                     0   

     SaleCondition_Normal  SaleCondition_Partial  
465                     1                      0  
924                     1                      0  
23                      1                      0  
422                     1                      0  
610                     1                      0  

[5 rows x 226 columns]

# Linear regression
    - MSE가 굉장히 형편업다. 모든데이터를 다 떼려박아서 그런건가
    

In [9]:
import sklearn.linear_model as lm
LR = lm.LinearRegression()
LR.fit(x_train,y_train)
print("r2 ", 1- LR.score(x_val,y_val))
from sklearn.metrics import mean_squared_error
mean_squared_error(LR.predict(x_val),y_val) # MSE 

r2  0.10796862995036094


608239923.1092814

# Lasso regression
    - 교차확인을 통해 hyperparameter 추정 하는 방법도 있지만 지금은 단순하게

In [10]:
lasso = lm.Lasso(alpha=0.01,max_iter=10000)
lasso.fit(x_train,y_train)
print("r2 ", 1- lasso.score(x_val,y_val))

# cv_lasso = lm.LassoCV(n_alphas = 100, fit_intercept = True, normalize = True)
# cv_lasso.fit(X_train, y_train)
# cv_lasso.alpha_

r2  0.10847011937492446


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)


# Ridge regression

In [61]:
ridge = lm.Ridge(alpha=0.01,solver= 'sag',tol=0.1)
ridge.fit(x_train,y_train)
print(ridge)
print("r2  ", 1- ridge.score(x_val,y_val))

mean_squared_error(ridge.predict(x_val),y_val) # MSE 
# cv_lasso = lm.LassoCV(n_alphas = 100, fit_intercept = True, normalize = True)
# cv_lasso.fit(X_train, y_train)
# cv_lasso.alpha_

Ridge(alpha=0.01, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='sag', tol=0.1)
r2   0.8243783801112157


4644125268.259221

# Linear Regression + Stepwise 선택법(hybrid)
    - 10.8% 에서 10.6% 로 소폭 상승했다. 
    - 제대로 확인하려면 Cross Validation이 필요하다
    - Cross Validation으로 한번 구해보자

In [13]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
def stepwise_selection(X, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True):
    """ Perform a forward-backward feature selection 
    based on p-value from statsmodels.api.OLS
    Arguments:
        X - pandas.DataFrame with candidate features
        y - list-like with the target
        initial_list - list of features to start with (column names of X)
        threshold_in - include a feature if its p-value < threshold_in
        threshold_out - exclude a feature if its p-value > threshold_out
        verbose - whether to print the sequence of inclusions and exclusions
    Returns: list of selected features 
    Always set threshold_in < threshold_out to avoid infinite looping.
    See https://en.wikipedia.org/wiki/Stepwise_regression for the details
    """
    included = list(initial_list)
    while True:
        changed=False
        # forward step
        excluded = list(set(X.columns)-set(included))
        new_pval = pd.Series(index=excluded)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min()
        if best_pval < threshold_in:
            best_feature = new_pval.argmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

        # backward step
        model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
        # use all coefs except intercept
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() # null if pvalues is empty
        if worst_pval > threshold_out:
            changed=True
            worst_feature = pvalues.argmax()
            included.remove(worst_feature)
            if verbose:
                print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
        if not changed:
            break
    return included

result = stepwise_selection(x_train, y_train)

# model = forward_selected(tra, 'SalePrice')

C:\ProgramData\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools
C:\ProgramData\Anaconda3\lib\site-packages\statsmodels\base\model.py:1036: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
C:\ProgramData\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
C:\ProgramData\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
C:\ProgramData\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:1821: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_l

Add  OverallQual                    with p-value 6.23446e-223
Add  GrLivArea                      with p-value 2.16276e-63
Add  BsmtFinSF1                     with p-value 1.1344e-27
Add  GarageCars                     with p-value 1.09649e-21
Add  MSSubClass                     with p-value 1.05039e-15
Add  Neighborhood_NridgHt           with p-value 5.24873e-10
Add  Neighborhood_StoneBr           with p-value 2.24889e-12
Add  Neighborhood_NoRidge           with p-value 3.15503e-11
Add  BsmtExposure_Gd                with p-value 2.61278e-10
Add  Condition2_PosN                with p-value 2.44782e-10
Add  YearRemodAdd                   with p-value 3.72366e-09
Add  RoofMatl_WdShngl               with p-value 1.14466e-09
Add  SaleType_New                   with p-value 2.85752e-06
Add  Neighborhood_Crawfor           with p-value 4.20035e-06
Add  Exterior1st_Stucco             with p-value 1.75489e-06
Add  LotConfig_CulDSac              with p-value 8.49022e-06
Add  RoofMatl_CompShg   

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:45: FutureWarning: 'argmax' is deprecated. Use 'idxmax' instead. The behavior of 'argmax' will be corrected to return the positional maximum in the future. Use 'series.values.argmax' to get the position of the maximum now.


Add  LotArea                        with p-value 1.7171e-05
Drop Exterior1st_Stucco             with p-value 0.10691
Add  YearBuilt                      with p-value 1.21954e-05
Add  OverallCond                    with p-value 9.25367e-09
Drop YearRemodAdd                   with p-value 0.149292
Add  BedroomAbvGr                   with p-value 3.17052e-07
Add  BsmtQual_Gd                    with p-value 1.74897e-06
Add  BsmtQual_TA                    with p-value 1.87236e-07
Add  TotalBsmtSF                    with p-value 6.09174e-05
Add  KitchenAbvGr                   with p-value 1.5525e-05
Add  PoolArea                       with p-value 0.000435133
Add  KitchenQual_Gd                 with p-value 0.00121916
Add  KitchenQual_TA                 with p-value 8.90505e-10
Add  Exterior1st_BrkFace            with p-value 0.000667678
Drop Fireplaces                     with p-value 0.0587392
Add  KitchenQual_Fa                 with p-value 0.00135361
Add  MasVnrArea                     w

In [16]:
# result feature
LR2 = lm.LinearRegression()
LR2.fit(x_train.loc[:,result],y_train)
print(' ', 1-LR2.score(x_val.loc[:,result],y_val))

Validation error :  0.10658277676297934


# Cross validation으로 model 평가

In [17]:
def rmse_cv(model,X,y):
    rmse = np.sqrt(-cross_val_score(model, X, y, scoring="neg_mean_squared_error", cv=5))
#     print(rmse)
    return rmse

In [19]:
from sklearn.model_selection import KFold, cross_val_score
kfold = KFold(n_splits = 10)
models = [lm.LinearRegression(),LR2,lasso,lm.Ridge()]

cv_results = []
# cross validation
for model in models :
        cv_results.append(rmse_cv(model,x_train.loc[:,result],y_train))
# np.sqrt(-cross_val_score(model, X, y, scoring="neg_mean_squared_error", cv=5))

In [46]:
from sklearn.metrics import mean_squared_error
# 결과
print("Linear regression : ", cv_results[0].mean(), cv_results[0].std())
print("Linear regression aft selection : ", cv_results[1].mean(), cv_results[1].std())
print("lasso regression : ", '{:.6f}'.format(cv_results[2].mean()), cv_results[2].std())
print("Ridge regression : ", cv_results[3].mean(), cv_results[3].std())

Linear regression :  41799.55290072597 20054.191595052842
Linear regression aft selection :  41799.55290072597 20054.191595052842
lasso regression :  41534.292188 19768.35901096833
Ridge regression :  36047.69273991686 14682.301993870466


In [ ]:
cv_results[0].mean()

In [ ]:
from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()
X_scaled = scaler.fit(x_train).transform(x_train)

In [ ]:
cv_results2 = []
# cross validation
for model in models :
        cv_results2.append(rmse_cv(model,X_scaled,y_train))

In [ ]:
from sklearn.metrics import mean_squared_error
# 결과
print("Linear regression : ", cv_results2[0].mean(), cv_results2[0].std())
print("Linear regression aft selection : ", cv_results2[1].mean(), cv_results2[1].std())
print("lasso regression : ", cv_results2[2].mean(), cv_results2[2].std())
print("Ridge regression : ", cv_results2[3].mean(), cv_results2[3].std())

# test set에 넣어서 제출 파일만들기


In [21]:
df = pd.read_csv('sample_submission.csv')
df.head()

Id      SalePrice
0  1461  169277.052498
1  1462  187758.393989
2  1463  183583.683570
3  1464  179317.477511
4  1465  150730.079977

In [36]:
full_data[1] = full_data[1].dropna(axis=0)
ridge = lm.Ridge(alpha=60)
ridge.fit(x_train.loc[:,result],y_train)
# print("r2  ", 1- ridge.score(x_val,y_val))

Ridge(alpha=60, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [37]:
ridge.predict(full_data[1].loc[:,result])

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  """Entry point for launching an IPython kernel.
C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py:1367: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self._getitem_tuple(key)


ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [45]:
full_data[1][full_data[0].isnull().sum()>0]

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match